targets/f7/furi_hal/furi_hal_region.c

```diff
@@ -71,66 +71,80 @@ const FuriHalRegion furi_hal_region_jp = {
             .duty_cycle = 50,
         }}};
 
-static const FuriHalRegion* furi_hal_region = NULL;
+static const FuriHalRegion* const furi_hal_static_regions[] = {
+    [FuriHalVersionRegionUnknown] = &furi_hal_region_zero,
+    [FuriHalVersionRegionEuRu] = &furi_hal_region_eu_ru,
+    [FuriHalVersionRegionUsCaAu] = &furi_hal_region_us_ca_au,
+    [FuriHalVersionRegionJp] = &furi_hal_region_jp,
+};
+
+static FuriHalRegion* furi_hal_dynamic_region;
+static FuriMutex* furi_hal_dynamic_region_mutex;
 
 void furi_hal_region_init(void) {
-    FuriHalVersionRegion region = furi_hal_version_get_hw_region();
-
-    if(region == FuriHalVersionRegionUnknown) {
-        furi_hal_region = &furi_hal_region_zero;
-    } else if(region == FuriHalVersionRegionEuRu) {
-        furi_hal_region = &furi_hal_region_eu_ru;
-    } else if(region == FuriHalVersionRegionUsCaAu) {
-        furi_hal_region = &furi_hal_region_us_ca_au;
-    } else if(region == FuriHalVersionRegionJp) {
-        furi_hal_region = &furi_hal_region_jp;
-    }
+    furi_assert(furi_hal_dynamic_region_mutex == NULL);
+    furi_hal_dynamic_region_mutex = furi_mutex_alloc(FuriMutexTypeNormal);
 }
 
 const FuriHalRegion* furi_hal_region_get(void) {
-    return furi_hal_region;
+    const FuriHalVersionRegion region = furi_hal_version_get_hw_region();
+    const FuriHalRegion* ret;
+
+    if(region < FuriHalVersionRegionWorld) {
+        ret = furi_hal_static_regions[region];
```

Comment:
wouldn't that completely prevent applying provisioned region for r00-r03 devices?

Reply:
fixed